Track UserGroup modifications with RealmAuditLog

PIG208 opened the pull request:
Fixes: #19528

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

timabbott commented:
@sahil839 can you review this one?

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -241,7 +271,19 @@ def create_system_user_groups_for_realm(realm: Realm) -> Dict[int, UserGroup]:
+                event_time=creation_time,
+                modified_user_group=user_group,
+            )
+            for user_group in system_user_groups_list
```
While we are at this, do we also want to incldue creating `GroupGroupMembership` objects in the atomic transaction? Though `create_system_groups_for_realm` is called inside a transaction from `do_create_realm` but not when calling from `do_impor_realm`.

PIG208 replied:
That makes sense. We don't want to end up in a bad state when the system groups are half-prepared. There are functions in `zerver.lib.user_groups` that I want to move to `zerver.actions.user_groups`. With a separate refactoring, we might wrap this entire function inside a transaction.

sahil839 reviewed (commented):
LGTM. I left one question too. I assume more commits will be added to handle subgroup updates and group deletion since this is a draft PR.

PIG208 commented:
> LGTM. I left one question too. I assume more commits will be added to handle subgroup updates and group deletion since this is a draft PR.

Yes. That's the plan. For subgroup membership updates, I'm planning to only audit the case when `modified_user_group` is the supergroup. When all the codepaths for user group changes are covered, we can add a migration to backfill audit logs.

timabbott commented:
@PIG208 I see this is still tagged as a draft. Let me know when I should review it; I think it's also worth considering whether there's some intermediate checkpoint that would be useful to integrate before everything is done.

PIG208 commented:
For this PR, I still have to hook up audit log generation for deleted user groups and test cases. At this point, we can get #23834 merged first, which is a series of refactorings to get `acting_user` available for `user_group` related helpers.

PIG208 committed:
migrations: Backfill audit logs for user groups.

This is quite similar to 0450_backfill_subscription_auditlogs. Here we
backfill for RealmAuditLog with event types USER_GROUP_CREATED,
USER_GROUP_DIRECT_USER_MEMBERSHIP_ADDED,
USER_GROUP_DIRECT_SUBGROUP_MEMBERSHIP_ADDED, and
USER_GROUP_DIRECT_SUPERGROUP_MEMBERSHIP_ADDED.

Fixes #19528.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 commented:
The PR has been updated. Haven't thoroughly tested out the migration yet. The first commit should be ready to merge. The rest of them should be in a good shape right now.

PIG208 commented:
The current plan for this is to track supergroup changes as well per [Alex's suggestion on CZO](https://chat.zulip.org/#narrow/stream/3-backend/topic/Audit.20log.20for.20UserGroup/near/1586720)

PIG208 commented:
Opened #25970 for the first part of the PR.

PIG208 commented:
The PR is updated to support super group changes. I also renamed the event types so it is obvious that they refer to direct memberships.

PIG208 commented:
No, these two PRs can be merged in either order. Updating this PR to accommodate the other migration would be quite straight-forward.

sahil839 commented on zerver/migrations/0454_backfill_usergroup_realmauditlogs.py:
```diff
@@ -0,0 +1,121 @@
+                        modified_user_group_id=user_group.id,
+                        event_type=USER_GROUP_DIRECT_SUBGROUP_MEMBERSHIP_ADDED,
+                    )
+                    .distinct("modified_user_group_id")
```
I think this is incorrect. We can have `RealmAuditLog` objects added for multiple subgroups of a user_group and using `distinct` would return only one of them.

PIG208 replied:
Thanks for catching that. I have removed this line and adjusted the test case to verify if it works properly.

sahil839 reviewed (commented):
Posted few comments. Looks good otherwise.

PIG208 commented on version.py:
```diff
@@ -48,4 +48,4 @@
 #   minor version bump suffices.
 
-PROVISION_VERSION = (243, 1)
+PROVISION_VERSION = (243, 2)
```
Oh, good point! I have a misconception the whole time that migrations need a PROVISION bump. Thanks for clearing that up.

PIG208 commented on zerver/actions/user_groups.py:
```diff
@@ -271,51 +271,73 @@ def do_send_subgroups_update_event(
+            extra_data=orjson.dumps({"supergroup_id": supergroup.id}).decode(),
+        )
+        for subgroup_id in subgroup_ids
+    )
```
Yeah, good idea.

PIG208 commented:
The PR has been updated.

sahil839 commented on zerver/actions/user_groups.py:
```diff
@@ -278,6 +279,19 @@ def add_subgroups_to_user_group(
+            event_type=RealmAuditLog.USER_GROUP_DIRECT_SUBGROUP_MEMBERSHIP_ADDED,
+            event_time=now,
+            acting_user=acting_user,
+            extra_data=orjson.dumps({"subgroup_id": subgroup_id}).decode(),
```
We can possible have `extra_data` store a list of subgroup IDs with `subgroup_ids` as key instead of different objects for each subgroup, but I am not sure what would be better. Would wait for feedback from Tim on this.

PIG208 replied:
Sidenote: perhaps it can be similar to what we did for the realm filter events, by having a complete set of subgroups there are at that point. But it might become very clumsy at some point.

sahil839 commented on zerver/tests/test_migrations.py:
```diff
@@ -0,0 +1,209 @@
+        iago = self.example_user("iago")
+        hamlet = self.example_user("hamlet")
+
+        # No non-backfilled entries are expected other than the 4 we populated
```
This comment should have "6" here instead of "4".

sahil839 commented on zerver/actions/user_groups.py:
```diff
@@ -271,51 +271,75 @@ def do_send_subgroups_update_event(
+                orjson.dumps({"supergroup_id": supergroup.id}).decode(),
+                subgroup_id,
+            ),
+        )
```
I am finding this difficult to read and similar for changes in `remove_subgroups_from_user_group`. Would probably be better to have something like what you have done in `create_system_user_groups_for_realm` -
```python
realmauditlog_objects = []
for subgroup in subgroup_ids:
    realmauditlog_objects.extend(
        [
            RealmAuditLog(
                realm=supergroup.realm,
                modified_user_group_id=supergroup.id,
                event_type=RealmAuditLog.USER_GROUP_DIRECT_SUBGROUP_MEMBERSHIP_REMOVED,
                event_time=now,
                acting_user=acting_user,
                extra_data=orjson.dumps({"subgroup_id": subgroup_id}).decode(),
            ),
             RealmAuditLog(
                realm=supergroup.realm,
                modified_user_group_id=subgroup_id,
                event_type=RealmAuditLog.USER_GROUP_DIRECT_SUPERGROUP_MEMBERSHIP_REMOVED,
                event_time=now,
                acting_user=acting_user,
                extra_data=orjson.dumps({"supergroup_id": supergroup.id}).decode(),
            )
        ]
    )
RealmAuditLog.objects.bulk_create(...)
```

sahil839 commented on zerver/lib/user_groups.py:
```diff
@@ -295,13 +296,25 @@ def create_system_user_groups_for_realm(realm: Realm) -> Dict[int, UserGroup]:
+        )
 
     GroupGroupMembership.objects.bulk_create(subgroup_objects)
+    RealmAuditLog.objects.bulk_create(realmauditlog_objects)
```
I think you forgot to do the changes here, `USER_GROUP_CREATED` are still created in a separate `bulk_create` call.

sahil839 reviewed (commented):
Posted a few more comments. Thanks!

PIG208 commented:
@sahil839 Thanks for the review! The PR is ready for another look.

alexmv commented on zerver/tests/test_audit_log.py:
```diff
@@ -1151,3 +1152,27 @@ def test_change_user_group_subgroups_memberships(self) -> None:
                 {"supergroup_id": user_group.id},
             )
+
+    def test_user_group_deletion(self) -> None:
+        hamlet = self.example_user("hamlet")
+        user_group = check_add_user_group(
+            hamlet.realm, "foo", [hamlet], description="Some description", acting_user=hamlet
+        )
+        now = timezone_now()
+        check_delete_user_group(user_group.id, acting_user=hamlet)
+        audit_log = RealmAuditLog.objects.filter(
+            realm=hamlet.realm,
+            event_time__gte=now,
+            event_type=RealmAuditLog.USER_GROUP_DELETED,
+        )
+        self.assert_length(audit_log, 1)
+        self.assertEqual(audit_log[0].acting_user, hamlet)
+        self.assertDictEqual(
+            orjson.loads(assert_is_not_none(audit_log[0].extra_data)),
+            {
+                "name": "foo",
+                "realm_id": hamlet.realm_id,
+                "description": "Some description",
+                "is_system_group": False,
+            },
+        )
```
Can we have a test for what happens with nested groups?  That is, if A is in B, and I delete B, does A get deleted?  Does it get an USER_GROUP_DELETED audit log entry?  I expect "yes" and "yes," but I suspect the code as written is "yes" and "no" respectively.

PIG208 replied:
Did you mean audit log entries for `USER_GROUP_DIRECT_SUPER/SUBGROUP_MEMBERSHIP_DELETED` event? I think we do a cascade delete only for the `GroupGroupMembership` rows when a related user group is deleted, but not for the user groups themselves. I think we do need to generate these audit log entries.

alexmv replied:
No, I mean what happens to subgroups when you delete the supergroup.

sahil839 replied:
The subgroup remains as it is when the supergroup is deleted and as mentioned above by Zixuan the `GroupGroupMembership` objects will be deleted.

alexmv replied:
Ah -- OK.  I guess it's necessary to leave the sub-group, particularly if we're allowing subgroup C to be in both group A and group B.  Makes sense!  But we do at least need the `USER_GROUP_DIRECT_SUPER` event.

alexmv commented on zerver/actions/user_groups.py:
```diff
@@ -347,9 +347,21 @@ def do_send_delete_user_group_event(realm: Realm, user_group_id: int, realm_id:
+        event_type=RealmAuditLog.USER_GROUP_DELETED,
+        event_time=timezone_now(),
+        extra_data=orjson.dumps(
+            {
+                "name": user_group.name,
+                "realm_id": user_group.realm_id,
+                "description": user_group.description,
+                "is_system_group": user_group.is_system_group,
+            }
```
It might be useful to list the set of members that were in it upon deletion, as well.

alexmv replied:
Ideally we take the lock as part of checking the access control.

alexmv commented on zerver/tests/test_migrations.py:
```diff
@@ -1,113 +0,0 @@
-# Django's database transaction model mean it does not super work to
-# have a lot of migrations tested in this file at once; so we usually
-# delete the old migration tests when adding a new one, so this file
-# always has a single migration test in it as an example.
```
Shouldn't we leave this file, unless we're substituting a new migration test?

alexmv replied:
Ah, I just got to the next commit.  You should squash this change into the subsequent commit.

PIG208 commented:
Aduitlog for user group deletion and the backfilling migration is blocked by #26048. But it will probably be helpful to get the rest of this PR merged while we are on #26048.

PIG208 commented:
@zulipbot add "blocked"

PIG208 commented:
@zulipbot remove "blocked"

PIG208 commented:
@zulipbot remove "blocked"

PIG208 committed:
user_groups: Add helper to find user group referrer in settings.

This maintains a hardcoded list of model attributes that define the
group permission settings dict. Using id_field_name for each setting in
each group permission settings dict, we can lookup the instances that
refer to the user group.

The performance of this query is proportional to the total number of settings
that may contain a reference to a user group. We may optimize it further
so that it takes up to one query per table, by checking all
"id_field_name"s at the same time.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
user_groups: Add basic support for deactivating user groups.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
user_groups: Add deactivated field to UserGroup.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 committed:
user_groups: Support user group reactivation.

Signed-off-by: Zixuan James Li <p359101898@gmail.com>

PIG208 commented:
@zulipbot remove "maintainer review"

sahil839 commented:
@PIG208 what is the status of this PR? I see the PRs on which this was blocked have been merged.

PIG208 commented:
If I remember correctly, what's left to be implemented is user group deactivation (to replace user group deletion) discussed [here](https://chat.zulip.org/#narrow/stream/101-design/topic/Support.20user.20group.20deactivation), before we can add the backfilling migration. The deactivation part has been a WIP.

sahil839 commented:
Ok, I will take a look at the discussion for deactivating groups and will see what we can do to complete it.
